### PR TITLE
NGINX Dynamic module with  NGINX Ingress controller

### PR DESCRIPTION
**Pull request by jasonwilliams14**
### Proposed changes
Doc describing how to build NGINX Ingress with a NGINX plus dynamic module.

### Checklist
Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [CONTRIBUTING](https://github.com/nginxinc/kubernetes-ingress/blob/main/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that all unit tests pass after adding my changes
- [x] I have updated necessary documentation
- [x] I have rebased my branch onto main
- [x] I will ensure my PR is targeting the main branch and pulling from my branch from my own fork

### CODE REVIEW

File: docs/content/tutorials/nginx-dynamic-module.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+  Use NGINX Ingress Controller with dynamic module
+weight: 1800
+doctypes: ["concept"]
+toc: true
+docs: "DOCS-1231"
+---
+
+## How to use a NGINX dynamic module with NGINX Ingress controller
+
+NGINX has several dnynamic modules that can be used to add additional features and capabilities. NGINX Ingress controller can also use these same NGINX modules with NGINX Ingress controller.
```
Comment (jjngx) [inline docs/content/tutorials/nginx-dynamic-module.md:13]: just a typo in `dnynamic`

File: docs/content/tutorials/nginx-dynamic-module.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+```
+config:
+  name: nginx-ingress
+  entries:
+    main-snippets: load_module modules/ngx_http_headers_more_filder_module.so;
+    http-snippets: underscores_in_headers on;
+    lb-method: "least_time last_byte"
+```
+
+kubectl exec -it -n nginx-ingress <nginx_ingress_pod> -- nginx -T
```
Comment (jjngx) [inline docs/content/tutorials/nginx-dynamic-module.md:113]: do we want this line to be formatted as a code/bash?

File: docs/content/tutorials/nginx-dynamic-module.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+Once you have the image cloned, edit the `Dockerfile` located in the `build` directory of the cloned repo.
+
+
+In our example, we are going to add the `Headers-more` dynamic module to our NGINX Ingress controller image.
+
+Locate your preferred OS version that you want to modify in the DOckerfile (debian, alpine etc.).
+Here is a snippet of the `Dockerfile` for `debian-plus`
+
+```
+FROM debian:${DEBIAN_VERSION} AS debian-plus
```
Comment (lucacome) [inline docs/content/tutorials/nginx-dynamic-module.md:44]:
```suggestion
FROM debian:11-slim AS debian-plus
```

File: docs/content/tutorials/nginx-dynamic-module.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+```
+git clone git@github.com:nginxinc/kubernetes-ingress.git
+```
+
+Once you have the image cloned, edit the `Dockerfile` located in the `build` directory of the cloned repo.
+
+
+In our example, we are going to add the `Headers-more` dynamic module to our NGINX Ingress controller image.
+
+Locate your preferred OS version that you want to modify in the DOckerfile (debian, alpine etc.).
```
Comment (lucacome) [inline docs/content/tutorials/nginx-dynamic-module.md:40]:
```suggestion
Locate your preferred OS version that you want to modify in the Dockerfile (debian, alpine etc.).
```

File: docs/content/tutorials/nginx-dynamic-module.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+title: NGINX Ingress Controller with dynamic module
+description: |
+  Use NGINX Ingress Controller with dynamic module
+weight: 1800
+doctypes: ["concept"]
+toc: true
+docs: "DOCS-1231"
+---
+
+## How to use a NGINX dynamic module with NGINX Ingress controller
```
Comment (lucacome) [inline docs/content/tutorials/nginx-dynamic-module.md:11]:
```suggestion
## How to use an NGINX dynamic module with F5 NGINX Ingress Controller
```

File: docs/content/tutorials/nginx-dynamic-module.md (hunk 1 above)
Comment (lucacome) [inline docs/content/tutorials/nginx-dynamic-module.md:13]:
```suggestion
NGINX has several dynamic modules that can be used to add additional features and capabilities. NGINX Ingress Controller can also use these same NGINX modules with NGINX Ingress Controller.
```

File: docs/content/tutorials/nginx-dynamic-module.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+NGINX has several dnynamic modules that can be used to add additional features and capabilities. NGINX Ingress controller can also use these same NGINX modules with NGINX Ingress controller.
+You will need to modify your NGINX Ingress controller build to ensure it is built into the image, and then load it into NGINX Ingress controller.
```
Comment (lucacome) [inline docs/content/tutorials/nginx-dynamic-module.md:14]:
```suggestion
You will need to modify your NGINX Ingress Controller build to ensure it is built into the image, and then load it into NGINX Ingress Controller.
```

File: docs/content/tutorials/nginx-dynamic-module.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+This document will walk you throw how to accomplish this task.
+
+
+You can find more NGINX dynamic modules on the NGINX Plus website:
+[NGINX Dynamic modules](https://docs.nginx.com/nginx/admin-guide/dynamic-modules/dynamic-modules/)
+
+Here are the basic steps that you will need to perform.
+
+- Update the Dockerfile and add your dynamic module you want to add to your NGINX Ingress controller image
```
Comment (lucacome) [inline docs/content/tutorials/nginx-dynamic-module.md:23]:
```suggestion
- Update the Dockerfile and add the dynamic module you want to add to your NGINX Ingress Controller image
```

File: docs/content/tutorials/nginx-dynamic-module.md [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+- Build the image with updated Dockerfile, reflecting the image you are goign to add to the image build.
+- Load the module using a `configmap` into NGINX Ingress controller
+
+In order to build a custom NGINX Ingress image with specific modules, the first step you need to take is to modify the `Dockerfile` located in the `./build` directory of the NGINXINC repo.
```
Comment (lucacome) [inline docs/content/tutorials/nginx-dynamic-module.md:27]:
```suggestion
In order to build a custom NGINX Ingress Controller image with specific modules, the first step you need to take is to modify the `Dockerfile` located in the `./build` directory of the NGINXINC repo.
```

File: docs/content/tutorials/nginx-dynamic-module.md [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+Clone the NGINX Ingress controller repo:
```
Comment (lucacome) [inline docs/content/tutorials/nginx-dynamic-module.md:29]:
```suggestion
Clone the NGINX Ingress Controller repo:
```

File: docs/content/tutorials/nginx-dynamic-module.md [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+```
+git clone git@github.com:nginxinc/kubernetes-ingress.git
+```
+
+Once you have the image cloned, edit the `Dockerfile` located in the `build` directory of the cloned repo.
+
+
+In our example, we are going to add the `Headers-more` dynamic module to our NGINX Ingress controller image.
```
Comment (lucacome) [inline docs/content/tutorials/nginx-dynamic-module.md:38]:
```suggestion
In our example, we are going to add the `Headers-more` dynamic module to our NGINX Ingress Controller image.
```

File: docs/content/tutorials/nginx-dynamic-module.md [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+	&& rm -rf /var/lib/apt/lists/*
+```
+
+Look for a line similar to the following line in the `Dockerfile`:
+
+```
+apt-get install --no-install-recommends --no-install-suggests -y nginx-plus nginx-plus-module-njs
+```
+
+This is the line you will want to modify/add the module you want to have loaded into NGINX Ingress controller.
```
Comment (lucacome) [inline docs/content/tutorials/nginx-dynamic-module.md:71]:
```suggestion
This is the line you will want to modify/add the module you want to have loaded into NGINX Ingress Controller.
```

File: docs/content/tutorials/nginx-dynamic-module.md [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+```
+apt-get install --no-install-recommends --no-install-suggests -y nginx-plus nginx-plus-module-njs nginx-plus-module-headers-more
+```
+
+In the above example, I added a single module line:
+```
+nginx-plus-module-headers-more
+```
+
+After the new NGINX Ingress module image has been built successfully, the next step is to load the module into your NGINX Ingress controller when it will be deployed into your Kubernetes cluster.
```
Comment (lucacome) [inline docs/content/tutorials/nginx-dynamic-module.md:83]:
```suggestion
After the new NGINX Ingress module image has been built successfully, the next step is to load the module into your NGINX Ingress Controller when it will be deployed into your Kubernetes cluster.
```

File: docs/content/tutorials/nginx-dynamic-module.md [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+apiVersion: v1
+metadata:
+  name: nginx-config
+  namespace: nginx-ingress
+data:
+  main-snippets: |
+    load_module modules/ngx_http_headers_more_filter_module.so;
+```
+
+With the above `configmap` configured to load in our `ngx_http_headers_more` module, NGINX Ingress controller will not load that module.
```
Comment (lucacome) [inline docs/content/tutorials/nginx-dynamic-module.md:99]:
```suggestion
With the above `configmap` configured to load in our `ngx_http_headers_more` module, NGINX Ingress Controller will not load that module.
```

File: docs/content/tutorials/nginx-dynamic-module.md [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+You can verify this be executing `nginx -T` in the NGINX Ingress controller pod:
```
Comment (lucacome) [inline docs/content/tutorials/nginx-dynamic-module.md:100]:
```suggestion
You can verify this be executing `nginx -T` in the NGINX Ingress Controller pod:
```

File: docs/content/tutorials/nginx-dynamic-module.md [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+  Use NGINX Ingress Controller with dynamic module
+weight: 1800
+doctypes: ["concept"]
+toc: true
+docs: "DOCS-1231"
+---
+
+## How to use an NGINX dynamic module with NGINX Ingress Controller
+
+NGINX has several dynamic modules that can be used to add additional features and capabilities. NGINX Ingress Controller can also use these same NGINX modules with NGINX Ingress Controller.
```
Comment (ADubhlaoich) [inline docs/content/tutorials/nginx-dynamic-module.md:13]:
```suggestion
NGINX has several dynamic modules that can be used to add additional features and capabilities. NGINX Ingress Controller can also use these same NGINX modules with NGINX Ingress Controller.
```

I would rephrase this to "additional features such as X, Y and Z": what exists right now ambi

File: docs/content/tutorials/nginx-dynamic-module.md [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+```yaml
+config:
+  name: nginx-ingress
+  entries:
+    main-snippets: load_module modules/ngx_http_headers_more_filder_module.so;
+    http-snippets: underscores_in_headers on;
+    lb-method: "least_time last_byte"
+```
+
+```console
```
Comment (ADubhlaoich) [inline docs/content/tutorials/nginx-dynamic-module.md:114]:
```suggestion
```shell
```

File: docs/content/tutorials/nginx-dynamic-module.md [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+	&& printf "%s\n" "deb https://pkgs.nginx.com/plus/${NGINX_PLUS_VERSION^^}/debian ${DEBIAN_VERSION} nginx-plus" > /etc/apt/sources.list.d/nginx-plus.list \
+	&& apt-get update \
+	&& apt-get install --no-install-recommends --no-install-suggests -y nginx-plus nginx-plus-module-njs \
+	&& apt-get purge --auto-remove -y apt-transport-https gnupg curl \
+	&& rm -rf /var/lib/apt/lists/*
+```
+
+Look for a line similar to the following line in the `Dockerfile`:
+
+```bash
```
Comment (ADubhlaoich) [inline docs/content/tutorials/nginx-dynamic-module.md:67]:
```suggestion
```shell
```

File: docs/content/tutorials/nginx-dynamic-module.md [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+apt-get install --no-install-recommends --no-install-suggests -y nginx-plus nginx-plus-module-njs
+```
+
+This is the line you will want to modify/add the module you want to have loaded into NGINX Ingress Controller.
+We are going to add the `headers-more` module. The updated line would look like this:
+
+```bash
```
Comment (ADubhlaoich) [inline docs/content/tutorials/nginx-dynamic-module.md:74]:
```suggestion
```shell
```

File: docs/content/tutorials/nginx-dynamic-module.md [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+
+NGINX has several dynamic modules that can be used to add additional features and capabilities. NGINX Ingress Controller can also use these same NGINX modules with NGINX Ingress Controller.
+You will need to modify your NGINX Ingress Controller build to ensure it is built into the image, and then load it into NGINX Ingress Controller.
+This document will walk you throw how to accomplish this task.
+
+
+You can find more NGINX dynamic modules on the NGINX Plus website:
+[NGINX Dynamic modules](https://docs.nginx.com/nginx/admin-guide/dynamic-modules/dynamic-modules/)
+
+Here are the basic steps that you will need to perform.
```
Comment (ADubhlaoich) [inline docs/content/tutorials/nginx-dynamic-module.md:21]:
```suggestion
Here are the basic steps you will need to perform.
```

**Review by ADubhlaoich — APPROVED** [review]
I pushed a lot of changes directly into this PR; they will need to be reviewed for accuracy, as I re-ordered certain examples for clarify while doing my best to keep the intent.

**Review by shaun-nx — APPROVED** [review]

**Review by jjngx — APPROVED** [review]
